File: 9_ICPRegistration/RobustICP.py
import open3d as o3d
import numpy as np
import matplotlib.pyplot as plt
import copy
import os
import sys
import time

def draw_registration_result(source, target, transformation):
    source_temp = copy.deepcopy(source)
    target_temp = copy.deepcopy(target)
    source_temp.paint_uniform_color([1, 0.706, 0])
    target_temp.paint_uniform_color([0, 0.651, 0.929])
    source_temp.transform(transformation)
    o3d.visualization.draw_geometries([source_temp, target_temp],
                                      zoom=0.4459,
                                      front=[0.9288, -0.2951, -0.2242],
                                      lookat=[1.6784, 2.0612, 1.4451],
                                      up=[-0.3402, -0.9189, -0.1996])

source = o3d.io.read_point_cloud("../test_data/ICP/cloud_bin_0.pcd")
target = o3d.io.read_point_cloud("../test_data/ICP/cloud_bin_1.pcd")
trans_init = np.asarray([[0.862, 0.011, -0.507, 0.5],
                         [-0.139, 0.967, -0.215, 0.7],
                         [0.487, 0.255, 0.835, -1.4], [0.0, 0.0, 0.0, 1.0]])
draw_registration_result(source, target, trans_init)


def apply_noise(pcd, mu, sigma):
    noisy_pcd = copy.deepcopy(pcd)
    points = np.asarray(noisy_pcd.points)
    points += np.random.normal(mu, sigma, size=points.shape)
    noisy_pcd.points = o3d.utility.Vector3dVector(points)
    return noisy_pcd

mu, sigma = 0, 0.1  # mean and standard deviation
source_noisy = apply_noise(source, mu, sigma)

print("Source PointCloud + noise:")
o3d.visualization.draw_geometries([source_noisy],
                                  zoom=0.4459,
                                  front=[0.353, -0.469, -0.809],
                                  lookat=[2.343, 2.217, 1.809],
                                  up=[-0.097, -0.879, 0.467])

threshold = 0.02
print("Vanilla point-to-plane ICP, threshold={}:".format(threshold))
p2l = o3d.pipelines.registration.TransformationEstimationPointToPlane()
reg_p2l = o3d.pipelines.registration.registration_icp(source_noisy, target,
                                                      threshold, trans_init,
                                                      p2l)

print(reg_p2l)
print("Transformation is:")
print(reg_p2l.transformation)
draw_registration_result(source, target, reg_p2l.transformation)


threshold = 1.0
print("Vanilla point-to-plane ICP, threshold={}:".format(threshold))
p2l = o3d.pipelines.registration.TransformationEstimationPointToPlane()
reg_p2l = o3d.pipelines.registration.registration_icp(source_noisy, target,
                                                      threshold, trans_init,
                                                      p2l)

print(reg_p2l)
print("Transformation is:")
print(reg_p2l.transformation)
draw_registration_result(source, target, reg_p2l.transformation)


print("Robust point-to-plane ICP, threshold={}:".format(threshold))
loss = o3d.pipelines.registration.TukeyLoss(k=sigma)
print("Using robust loss:", loss)
p2l = o3d.pipelines.registration.TransformationEstimationPointToPlane(loss)
reg_p2l = o3d.pipelines.registration.registration_icp(source_noisy, target,
                                                      threshold, trans_init,
                                                      p2l)
print(reg_p2l)
print("Transformation is:")
print(reg_p2l.transformation)
draw_registration_result(source, target, reg_p2l.transformation)

File: 9_ICPRegistration/src/RobustICP.cpp
#include <iostream>
#include <memory>
#include <thread>
#include <random>
#include <cstdlib>

#include <Eigen/Dense>

#include "open3d/Open3D.h"

void draw_registration_result(const open3d::geometry::PointCloud &source,
                              const open3d::geometry::PointCloud &target,
                              const Eigen::Matrix4d &Transforamtion){
    std::shared_ptr<open3d::geometry::PointCloud> source_temp(new open3d::geometry::PointCloud);
    std::shared_ptr<open3d::geometry::PointCloud> target_temp(new open3d::geometry::PointCloud);
    *source_temp = source;
    *target_temp = target;
    source_temp -> PaintUniformColor(Eigen::Vector3d(1, 0.706, 0));
    target_temp -> PaintUniformColor(Eigen::Vector3d(0, 0.651, 0.929));
    source_temp -> Transform(Transforamtion);
    open3d::visualization::DrawGeometries({source_temp, target_temp},
                                  "Registration result");
}

auto apply_noise(const open3d::geometry::PointCloud &pcd,
                 const double &mu, const double &sigma){
    auto noisy_pcd = std::make_shared<open3d::geometry::PointCloud>();
    *noisy_pcd = pcd;
    std::vector<Eigen::Vector3d> points = noisy_pcd -> points_;
    // Add Gaussian noise:
    for(size_t i=0; i < points.size(); i++){
        Eigen::Vector3d random_noise;
        random_noise << (mu + (double)sigma*(-1 + 2 * (rand()/(double)(RAND_MAX)))), 
                        (mu + (double)sigma*(-1 + 2 * (rand()/(double)(RAND_MAX)))),
                        (mu + (double)sigma*(-1 + 2 * (rand()/(double)(RAND_MAX))));
        points[i]= points[i] + random_noise;
    }
    noisy_pcd -> points_ = points;
    return noisy_pcd;
}

int main(int argc, char* argv[]) 
{
    auto source = std::make_shared<open3d::geometry::PointCloud>();
    source = open3d::io::CreatePointCloudFromFile("../../test_data/ICP/cloud_bin_0.pcd");
    auto target = std::make_shared<open3d::geometry::PointCloud>();
    target = open3d::io::CreatePointCloudFromFile("../../test_data/ICP/cloud_bin_1.pcd");
    double threshold = 0.02;
    // Manually set the transformation matrix
    Eigen::Matrix4d trans_init;
    trans_init << 0.862,  0.011, -0.507,  0.5,
                 -0.139,  0.967, -0.215,  0.7,
                  0.487,  0.255,  0.835, -1.4,
                  0.0,    0.0,    0.0,    1.0;
    std::cout << "Trans_init:" << std::endl;
    //draw_registration_result(*source, *target, trans_init);

    std::cout << "--------------------------------------------------" << std::endl; 
    std::cout << "Source Pointcloud + noise:" << std::endl;
    double mu = 0;
    double sigma = 0.1;
    auto source_noisy = apply_noise(*source, mu, sigma);
    open3d::visualization::DrawGeometries({source_noisy}, "Noisy PointCloud");
    
    std::cout << "--------------------------------------------------" << std::endl;
    double ICP_threshold = 0.02;
    std::cout << "Vanilla point-to-plane ICP, threshold= " << ICP_threshold << std::endl;
    auto reg_p2p = open3d::pipelines::registration::RegistrationICP(*source, *target, ICP_threshold,
                trans_init, open3d::pipelines::registration::TransformationEstimationPointToPoint());
    auto evaluation = open3d::pipelines::registration::EvaluateRegistration(*source_noisy, *target, ICP_threshold, reg_p2p.transformation_);
    std::cout << "  fitness: " << evaluation.fitness_  << " (The higher, the better)"<< std::endl;
    std::cout << "  inlier_rms: " << evaluation.inlier_rmse_  << " (The lower, the better)"<< std::endl;
    draw_registration_result(*source, *target, reg_p2p.transformation_);

    std::cout << "--------------------------------------------------" << std::endl;
    std::cout << "Robust point-to-plane ICP, threshold= " << ICP_threshold << std::endl;
    auto loss = open3d::pipelines::registration::TukeyLoss::TukeyLoss(0.1);  // ????
    // auto reg_p2l = open3d::pipelines::registration::RegistrationICP(*source, *target, ICP_threshold,
    //             trans_init, open3d::pipelines::registration::TransformationEstimationPointToPoint(
    //                 open3d::pipelines::registration::TukeyLoss::TukeyLoss(0.02)));

    return 0;
}